subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java

```diff
@@ -596,6 +596,24 @@ public int compareTo(Project otherProject) {
         return ProjectOrderingUtil.compare(this, otherProject);
     }
 
+    @Override
+    public int hashCode() {
+        return super.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object otherProject) {
+        if (otherProject == null) {
+            return false;
+        }
+        if (otherProject instanceof MutableStateAccessAwareProject) {
```

Comment:
🤔 It doesn't feel right for `DefaultProject` to know about `MutableStateAccessAwareProejct`

Reply:
Hm, I remember that we were even considering put the logic of `MutableStateAccessAwareProject` to `DefaultProject` when discussions about the implementation were going. So given that, and that both types belongs to the same module and even package(which is a bit overinflated, tho), for me it feels ok to have this connection.

Do you see a ways to fix `rawProject.equals(wrappedProject)` scenario without mentioning wrappedProject type in `DefaultProject`?

I mean, I see the ways with introducing some equality helpers and move the logic there (like `ProjectOrderingUtils`), but it doesn't feel conceptually different.

Reply:
@bamboo please continue the review in https://github.com/gradle/gradle/pull/30319